init/tee/kbs/kbs_types.c

```diff
@@ -50,3 +67,354 @@ kbs_challenge(CURL *curl, char *url, char *json_request, char *nonce)
 
         return 0;
 }
+
+/*
+ * Send all required materials (attestation report, certificate chain, etc..)
+ * to the attestation server for attestation.
+ */
+int
+kbs_attest(CURL *curl, char *url, struct snp_report *report, uint8_t *certs,
+                size_t certs_size, BIGNUM *mod, BIGNUM *exp)
+{
+        int rc;
+        char *json, errmsg[200];
+
+        rc = -1;
+        json = (char *) malloc(0x7000);
+        if (json == NULL) {
+                printf("ERROR: unable to allocate JSON buffer\n");
+
+                return rc;
+        }
+
+        /*
+         * Marshal the kbs_types Attestation JSON struct with the given
+         * attestation report and certificate chain.
+         */
+        kbs_attestation_marshal(report, certs, certs_size, json, mod, exp);
+
+        /*
+         * Ensure the error messaging string is empty, because we will
+         * eventually read this string as indicator of a cURL attestation
+         * server error.
+         */
+        strcpy(errmsg, "");
+
+        if (kbs_curl_post(curl, url, json, errmsg, KBS_CURL_ATTEST) < 0) {
+                printf("ERROR: could not complete KBS attestation\n");
+
+                rc = -1;
+                goto out;
+        }
+
+        /*
+         * If there is no error message, it can be assumed that the attestation
+         * was completed successfully.
+         */
+        if (strcmp(errmsg, "") != 0) {
+                rc = -1;
+                printf("ATTESTATION ERROR: %s\n", errmsg);
+
+                goto out;
+        }
+
+        rc = 0;
+
+out:
+        free((void *) json);
+
+        return rc;
+}
+
+/*
+ * Marshal a JSON string of the kbs_types Attestation struct from the given
+ * attestation report and certificate data.
+ */
+static void
+kbs_attestation_marshal(struct snp_report *report, uint8_t *certs,
+        size_t certs_size, char *json, BIGNUM *mod, BIGNUM *exp)
+{
+        char buf[4096];
+
+        sprintf(buf, "{");
+        strcpy(json, buf);
+
+        kbs_attestation_marshal_tee_pubkey(json, mod, exp);
+
+        sprintf(buf, "\"tee-evidence\":\"{\\\"report\\\":\\\"{");
```

Comment:
Is it worth marshaling  the entire report into JSON? Seems quite expensive to me when compared to, for instance, treating the entire report as a binary blob, encoding it in an hexadecimal string and sending it to the Attestation Server. Then the AS can decode the hexadecimal string into a `[u8]` array, and transmute it into a `sev::firmware::guest::types::AttestationReport`.

Reply:
Hex encoded in [1272e95](https://github.com/containers/libkrun/pull/108/commits/1272e956cf212f3877b3eb01b1f4d66719a7f8c7)